File: 23_climbing_Stairs/climb.cpp
#include <iostream>
#include <vector>

using namespace std;

// Climbing Stairs - Interview Question -
// Asked By Apple and Adobe
// in C++

int climbingStairs(int n) {
    vector<int> r;
    // Initial Result Set
    r.push_back(1);
    r.push_back(1);
    r.push_back(2);

    if (n <= 2)
        return r.at(n);

    for (int step = 3; step < n + 1; step++)
        r.push_back(r.at(step - 1) + r.at(step - 2));

    return r.at(n);
}

int main() {
    int n = 26; // Steps
    int ways = climbingStairs(n);
    cout << ways << endl; // 196418
}

File: 23_climbing_Stairs/climb.go
package main

import "fmt"

// Climbing Stairs - Interview Question -
// Asked By Apple and Adobe
// in Go

func climbingStairs(n int) int {
	r := []int{1, 1, 2} // Initial Result Set

	if n <= 2 {
		return r[n]
	}

	for step := 3; step < n+1; step++ {
		r = append(r, r[step-1]+r[step-2])
	}
	return r[n]
}

func main() {
	n := 26 // Steps
	ways := climbingStairs(n)
	fmt.Println(ways) // 196418
}
